etcdserver/server.go

```diff
@@ -1611,6 +1616,51 @@ func (s *EtcdServer) RemoveMember(ctx context.Context, id uint64) ([]*membership
 	return s.configure(ctx, cc)
 }
 
+// PromoteMember promotes a learner node to a voting node.
+func (s *EtcdServer) PromoteMember(ctx context.Context, id uint64) ([]*membership.Member, error) {
+	if err := s.checkMembershipOperationPermission(ctx); err != nil {
+		return nil, err
+	}
+
+	// check if we can promote this learner
+	if err := s.mayPromoteMember(types.ID(id)); err != nil {
+		return nil, err
+	}
+
+	// build the context for the promote confChange. mark IsLearner to false and IsPromote to true.
+	promoteChangeContext := membership.ConfigChangeContext{
+		Member: membership.Member{
+			ID: types.ID(id),
+		},
+		IsPromote: true,
+	}
+
+	b, err := json.Marshal(promoteChangeContext)
+	if err != nil {
+		return nil, err
+	}
+
+	cc := raftpb.ConfChange{
+		Type:    raftpb.ConfChangeAddNode,
+		NodeID:  id,
+		Context: b,
+	}
+
+	return s.configure(ctx, cc)
+}
+
+func (s *EtcdServer) mayPromoteMember(id types.ID) error {
+	if !s.Cfg.StrictReconfigCheck {
+		return nil
+	}
+	// TODO add more checks whether the member can be promoted.
+	// like learner progress check or if cluster is ready to promote a learner
+	// this is an example to get progress
+	fmt.Printf("raftStatus, %#v\n", raftStatus())
```

Comment:
we should not simply use fmt.printf as logging.

Reply:
This part is unfinished. This line should be removed for now. The actual implementation is in later commit (not included in part 2 PR): https://github.com/etcd-io/etcd/pull/10645/commits/dc79587cd1b05007516544db0bb576b70436aca3

Reply:
but checking this before the apply phase is ok.

Reply:
> but checking this before the apply phase is ok.

Not sure I understand. The member promote request should be rejected (if the reason is learner not ready) before it is appended to raft log. I think it is too late if we check learner progress in apply stage. In other words, applying a particular conf change raft entry should not depend on the progress of learner node in cluster. Some node may choose to apply this entry, whereas some node may not - depending on when they started applying this entry.

Reply:
yes, that is what i mean. checking this at the apply phase will not work since the cluster might not have a consistent view on progress unless the leader keep on propagating this information.

Reply:
> This part is unfinished. This line should be removed for now. The actual implementation is in later commit (not included in part 2 PR):

can you just make this a noop or a TODO panic?

Reply:
Currently it is noop - this checking function always return nil.

Reply:
it will print something out, can we remove that.

Reply:
Yes this print should be removed.